File: lib/features/ielts/domain/ielts_scoring.dart
/// 0–100 → IELTS band conversion. Keep in lock-step with `backend/app.py` and
/// `koras-web/src/lib/ielts/ielts-scoring.ts`. See 14.
double scoreToBand(num s) {
  if (s >= 92) return 9.0;
  if (s >= 87) return 8.5;
  if (s >= 82) return 8.0;
  if (s >= 77) return 7.5;
  if (s >= 72) return 7.0;
  if (s >= 65) return 6.5;
  if (s >= 58) return 6.0;
  if (s >= 52) return 5.5;
  if (s >= 45) return 5.0;
  if (s >= 38) return 4.5;
  return 4.0;
}

double roundToNearestHalfBand(num v) => (v * 2).round() / 2;

const ieltsCriterionLabels = <String, String>{
  'fluencyCoherence': 'Fluency & Coherence',
  'lexicalResource': 'Vocabulary',
  'grammarRangeAccuracy': 'Grammar',
  'pronunciation': 'Pronunciation',
};

File: supabase/functions/_shared/ielts-scoring.ts
// IELTS scoring helpers, ported from `koras-web/src/lib/ielts/ielts-scoring.ts`
// and the criteria labels from `ielts-types.ts`.

export type IELTSCriterionKey =
  | "fluencyCoherence"
  | "lexicalResource"
  | "grammarRangeAccuracy"
  | "pronunciation";

export const IELTS_CRITERIA_LABELS: Record<IELTSCriterionKey, string> = {
  fluencyCoherence: "Fluency & Coherence",
  lexicalResource: "Vocabulary",
  grammarRangeAccuracy: "Grammar",
  pronunciation: "Pronunciation",
};

export function roundToNearestHalfBand(value: number): number {
  return Math.round(value * 2) / 2;
}

export function strongestAndWeakest(
  scores: Record<string, number>,
): { strongest: IELTSCriterionKey; weakest: IELTSCriterionKey } {
  const entries = Object.entries(scores) as Array<[IELTSCriterionKey, number]>;
  let strongest = entries[0][0];
  let weakest = entries[0][0];
  for (const [key, value] of entries) {
    if (value > scores[strongest]) strongest = key;
    if (value < scores[weakest]) weakest = key;
  }
  return { strongest, weakest };
}
